rllib/agents/ppo/appo_torch_policy.py

```diff
@@ -8,391 +8,398 @@
+from ray.rllib.agents.ppo.appo_tf_policy import MakeAPPOModel
+from ray.rllib.evaluation.episode import Episode
+from ray.rllib.evaluation.postprocessing import (
+    compute_gae_for_sample_batch,
+    Postprocessing,
+)
+from ray.rllib.models.action_dist import ActionDistribution
 from ray.rllib.models.modelv2 import ModelV2
 from ray.rllib.models.torch.torch_action_dist import (
     TorchDistributionWrapper,
     TorchCategorical,
 )
-from ray.rllib.policy.policy import Policy
-from ray.rllib.policy.policy_template import build_policy_class
+from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
 from ray.rllib.policy.sample_batch import SampleBatch
 from ray.rllib.policy.torch_mixins import (
     EntropyCoeffSchedule,
+    GradClippingMixin,
     LearningRateSchedule,
+    KLCoeffMixin,
     ValueNetworkMixin,
+    TargetNetworkMixin,
 )
+from ray.rllib.policy.torch_policy_v2 import TorchPolicyV2
+from ray.rllib.utils.annotations import override
 from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.numpy import convert_to_numpy
 from ray.rllib.utils.torch_utils import (
-    apply_grad_clipping,
     explained_variance,
     global_norm,
     sequence_mask,
 )
-from ray.rllib.utils.typing import TensorType, TrainerConfigDict
+from ray.rllib.utils.typing import TensorType
 
 torch, nn = try_import_torch()
 
 logger = logging.getLogger(__name__)
 
 
-def appo_surrogate_loss(
-    policy: Policy,
-    model: ModelV2,
-    dist_class: Type[TorchDistributionWrapper],
-    train_batch: SampleBatch,
-) -> TensorType:
-    """Constructs the loss for APPO.
-
-    With IS modifications and V-trace for Advantage Estimation.
-
-    Args:
-        policy (Policy): The Policy to calculate the loss for.
-        model (ModelV2): The Model to calculate the loss for.
-        dist_class (Type[ActionDistribution]): The action distr. class.
-        train_batch (SampleBatch): The training data.
-
-    Returns:
-        Union[TensorType, List[TensorType]]: A single loss tensor or a list
-            of loss tensors.
-    """
-    target_model = policy.target_models[model]
-
-    model_out, _ = model(train_batch)
-    action_dist = dist_class(model_out, model)
-
-    if isinstance(policy.action_space, gym.spaces.Discrete):
-        is_multidiscrete = False
-        output_hidden_shape = [policy.action_space.n]
-    elif isinstance(policy.action_space, gym.spaces.multi_discrete.MultiDiscrete):
-        is_multidiscrete = True
-        output_hidden_shape = policy.action_space.nvec.astype(np.int32)
-    else:
-        is_multidiscrete = False
-        output_hidden_shape = 1
-
-    def _make_time_major(*args, **kwargs):
-        return make_time_major(
-            policy, train_batch.get(SampleBatch.SEQ_LENS), *args, **kwargs
+# We need this builder function because we want to share the same
+# custom logics between TF1 dynamic and TF2 eager policies.
+class APPOTorchPolicy(
+    MakeAPPOModel,
```

Comment:
Ah, I see: Cross-framework compatibility :) 

Ok, maybe define the utility function behind this in `appo.py` directly? B/c I feel like if it's truly framework agnostic, it shouldn't be in the tf-policy file either.

Reply:
I am keeping this in appo_tf_policy.py, since policy shouldn't import trainer.py either.
maybe add an appo_util.py or something ...

Reply:
Hmhm, importing the trainer module from policy works if we do the import in-line.